tls/s2n_resume.c

```diff
@@ -769,187 +769,147 @@ struct s2n_ticket_key *s2n_find_ticket_key(struct s2n_config *config, const uint
-    POSIX_GUARD(s2n_stuffer_write(to, &iv));
+    /* Initialize Additional Authenticated Data */
+    uint8_t aad_data[S2N_TICKET_AAD_LEN] = { 0 };
+    struct s2n_blob aad_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&aad_blob, aad_data, sizeof(aad_data)));
+    struct s2n_stuffer aad = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&aad, &aad_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->implicit_aad, S2N_TICKET_AAD_IMPLICIT_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->key_name, S2N_TICKET_KEY_NAME_LEN));
 
-    POSIX_GUARD(s2n_blob_init(&aes_key_blob, key->aes_key, S2N_AES256_KEY_LEN));
-    POSIX_GUARD(s2n_session_key_alloc(&aes_ticket_key));
-    POSIX_GUARD_RESULT(s2n_aes256_gcm.init(&aes_ticket_key));
-    POSIX_GUARD_RESULT(s2n_aes256_gcm.set_encryption_key(&aes_ticket_key, &aes_key_blob));
+    /* Write key name */
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(to, key->key_name, S2N_TICKET_KEY_NAME_LEN));
 
-    POSIX_GUARD(s2n_stuffer_init(&aad, &aad_blob));
-    POSIX_GUARD(s2n_stuffer_write_bytes(&aad, key->implicit_aad, S2N_TICKET_AAD_IMPLICIT_LEN));
-    POSIX_GUARD(s2n_stuffer_write_bytes(&aad, key->key_name, S2N_TICKET_KEY_NAME_LEN));
+    /* Write IV */
+    uint8_t iv_data[S2N_TLS_GCM_IV_LEN] = { 0 };
+    struct s2n_blob iv = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&iv, iv_data, sizeof(iv_data)));
+    RESULT_GUARD(s2n_get_public_random_data(&iv));
+    RESULT_GUARD_POSIX(s2n_stuffer_write(to, &iv));
 
-    uint32_t plaintext_header_size = s2n_stuffer_data_available(to);
-    POSIX_GUARD_RESULT(s2n_serialize_resumption_state(conn, to));
-    POSIX_GUARD(s2n_stuffer_skip_write(to, S2N_TLS_GCM_TAG_LEN));
+    /* Write serialized session state */
+    uint32_t plaintext_state_size = s2n_stuffer_data_available(to);
+    RESULT_GUARD(s2n_serialize_resumption_state(conn, to));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(to, S2N_TLS_GCM_TAG_LEN));
 
+    /* Initialize blob to be encrypted */
     struct s2n_blob state_blob = { 0 };
     struct s2n_stuffer copy_for_encryption = *to;
-    POSIX_GUARD(s2n_stuffer_skip_read(&copy_for_encryption, plaintext_header_size));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_read(&copy_for_encryption, plaintext_state_size));
     uint32_t state_blob_size = s2n_stuffer_data_available(&copy_for_encryption);
     uint8_t *state_blob_data = s2n_stuffer_raw_read(&copy_for_encryption, state_blob_size);
-    POSIX_ENSURE_REF(state_blob_data);
-    POSIX_GUARD(s2n_blob_init(&state_blob, state_blob_data, state_blob_size));
+    RESULT_ENSURE_REF(state_blob_data);
+    RESULT_GUARD_POSIX(s2n_blob_init(&state_blob, state_blob_data, state_blob_size));
 
-    POSIX_GUARD(s2n_aes256_gcm.io.aead.encrypt(&aes_ticket_key, &iv, &aad_blob, &state_blob, &state_blob));
+    RESULT_GUARD_POSIX(s2n_aes256_gcm.io.aead.encrypt(&aes_ticket_key, &iv, &aad_blob, &state_blob, &state_blob));
+    RESULT_GUARD(s2n_aes256_gcm.destroy_key(&aes_ticket_key));
 
```

Comment:
What's the difference between destroy_key and s2n_session_key_free? Do we need to do both on cleanup?

Reply:
destroy_key calls EVP_CIPHER_CTX_cleanup. s2n_session_key_free calls EVP_CIPHER_CTX_free(openssl) or EVP_AEAD_CTX_free(awslc).

EVP_CIPHER_CTX_free is the replacement for EVP_CIPHER_CTX_cleanup. If you look at current openssl code they both just call EVP_CIPHER_CTX_reset. So it should be fine to remove the destroy_key call.

Reply:
> If you look at current openssl code they both just call EVP_CIPHER_CTX_reset

Do they behave differently in earlier versions? Remember, we still need to support 1.0.2 :(

Reply:
[In 1.0.2](https://github.com/openssl/openssl/blob/OpenSSL_1_0_2/crypto/evp/evp_enc.c#L547), EVP_CIPHER_CTX_free is a wrapper around EVP_CIPHER_CTX_cleanup. So even in 1.0.2, it wouldn't be a change in behavior to switch to calling EVP_CIPHER_CTX_free().

What I can see from git blame is that EVP_CIPHER_CTX_init used to be paired with EVP_CIPHER_CTX_cleanup. Then, Openssl made the EVP_CIPHER_CTX struct opaque. So then they introduced EVP_CIPHER_CTX_new to be paired with EVP_CIPHER_CTX_free. These new functions internally allocate and deallocate cipher ctx memory.

In our library we just decided to call both of these function pairs? I really can't explain our current set up. It makes no sense to me. But regardless, I still think its fine to only call EVP_CIPHER_CTX_free since it's a wrapper around EVP_CIPHER_CTX_cleanup.